src/helper.rs

```diff
@@ -644,3 +645,114 @@ pub fn concat(xs: Expr) -> Expr {
+/// depending on if it is true (`if_true`) or false (`if_false`)
+#[inline]
+pub fn expr_if_else(scrutinee: Expr, if_true: Expr, if_false: Expr) -> Expr {
+    expr_match(scrutinee, [(Pattern::Bool(true), if_true), (Pattern::Bool(false), if_false)])
+}
+
+/// Helper function simulating the `bool::then` function that returns either `Some(if_true)` or `None` according
+/// to the boolean value of `scrutinee`.
+///
+/// Note that `if_true` is interpreted as having the parametric type of the resulting `Option<T>`. If the value
+/// we want to return is already an Option, use [`expr_if_else`] instead.
+#[inline]
+pub fn expr_opt_if(scrutinee: Expr, if_true: Expr) -> Expr {
+    expr_if_else(scrutinee, expr_some(if_true), expr_none())
+}
+
+/// Performs a table lookup operation over a sequence of entries `seq` of type `elem_type`.
+///
+/// Uses `f_getkey` to map from each entry to its key, and `query_key` as the query key.
+///
+/// Even if mulitple matching entries exist, only the first one is returned, as an Option.
+pub fn table_find(
+    seq: Expr,
+    elem_type: ValueType,
+    f_getkey: impl Fn(Expr) -> Expr,
+    query_key: Expr,
+) -> Expr {
+    let match0_or1 = flat_map_list(
+        lambda(
+            "list-entry",
+            expr_match(
+                seq_length(tuple_proj(var("list-entry"), 0)),
+                [
+                    (Pattern::Int(Bounds::at_least(1)), Expr::Seq(Vec::new())),
+                    (
+                        Pattern::Wildcard,
+                        expr_match(
+                            expr_eq(f_getkey(tuple_proj(var("list-entry"), 1)), query_key),
+                            [
+                                (Pattern::Bool(true), Expr::Seq(vec![var("entry")])),
+                                (Pattern::Bool(false), Expr::Seq(Vec::new())),
+                            ],
+                        ),
+                    ),
+                ],
+            ),
+        ),
+        elem_type,
+        seq,
+    );
+    seq_to_opt(match0_or1)
+}
+
+/// Performs the natural transformation from a singleton-or-empty `Expr::Seq` into `Option`.
+///
+/// # Notes
+///
+/// Not well-formed (i.e. will result in a runtime error) if the provided Expr can ever be a sequence of length 2 or higher.
+pub fn seq_to_opt(empty_or_singleton: Expr) -> Expr {
+    expr_match(
+        empty_or_singleton,
+        [
+            (Pattern::Seq(vec![bind("x")]), expr_some(var("x"))),
+            (Pattern::Seq(Vec::new()), expr_none()),
+        ],
+    )
+}
+
+/// Performs a fallible destructuring of the provided `expr` within the Format layer,
+/// either extracting the contents of a `Some(_)` value, or poisoning the current parse-state
+/// via `Format::Fail` if it is `None`.
+pub fn fmt_unwrap(expr: Expr) -> Format {
+    Format::Match(expr, vec![(pat_some(bind("x")), Format::Compute(var("x"))), (pat_none(), Format::Fail)])
+}
```

Comment:
While this is relatively easy to erase using `Format::LetFormat` or its helper `chain(F0, Label, F1)`, we might want to think about whether there is a 'good way' to produce a `Fail`-like behavior within the pure Expr layer, possibly restricted to `Compute(X)`.

Reply:
E.g.

```rust
pub enum Expr {
    /* everything we already have */
   Undefined,
}
```

where

$$
\mathrm{Decode}(\mathtt{Format::Compute}(\mathtt{Expr::Undefined})) \simeq \mathrm{Decode}(\mathtt{Format::Fail})
$$

Reply:
This might completely gum up the model though, so while forcing us to lift the unwrap into the Format layer is a bit awkward, it might be an easier approach overall.

Reply:
Non-terminating expressions do raise questions about order of evaluation etc. but arguably so does division by zero.